Version 1.0

Felix754 opened the pull request:
Pull request for receiving feedback from you🐦

Felix754 committed:
Version 1.0

ArtemiiH commented on decoder_output.txt:
```diff
@@ -0,0 +1,30 @@
+File Opened: 20150310.00.W.dwa_griby.grib
```
What is the purpose of storing output file in git. I don't think it is needed and it is better to remove it

ArtemiiH commented on main.cpp:
```diff
@@ -0,0 +1,216 @@
+#include <fstream>
+using namespace std;
+
+string fileName = "20150310.00.W.dwa_griby.grib";
+fstream file(fileName, ios::in | ios::binary);
```
It would be amazing if you could transform this simple script into a CLI programm, that can accept input and output file names.

ArtemiiH commented on main.cpp:
```diff
@@ -0,0 +1,216 @@
+        return 1;
+    }
+
+    ofstream fout("decoder_output.txt");
+    if (!fout)
+    {
+        return 1;
+    }
```
Would be nice to have some output for the user in this case as well, I guess,

ArtemiiH commented on main.cpp:
```diff
@@ -0,0 +1,216 @@
+
+    fout << "File Opened: " << fileName << endl;
+
+    // ------------------ Section 0 ------------------
```
General approach in a case like this - if you have "sections" of the code split by comment, you should move it to a function. Something that would get file, skip "GRIB" and return, and then next function, that would read and return length and version of the grip file

ArtemiiH commented on main.cpp:
```diff
@@ -0,0 +1,216 @@
+    streampos section0_start = file.tellg(); // Remember the position after "GRIB"
+
+    file.read(buffer, 4);
+    int length_of_grib_file = (unsigned char)buffer[0] << 16 |
+                              (unsigned char)buffer[1] << 8 |
+                              (unsigned char)buffer[2];
+    int version_of_grib = (unsigned char)buffer[3];
```
When you will refactor all the sections to separate functions, you should also create structures (or data classes, but I am not sure if there is such separate type of classes in C++) to hold relevant data that those functions will return.

ArtemiiH commented on main.cpp:
```diff
@@ -0,0 +1,216 @@
+        scale_factor_D = -scale_factor_D;
+    fout << "Scale factor D: " << scale_factor_D << endl;
+
+    // ------------------ Section 2 (GDS) ------------------
```
Same as above - move all sections into separate functions

ArtemiiH commented on main.cpp:
```diff
@@ -0,0 +1,216 @@
+    // Move to the next section
+    file.seekg(section2_start + static_cast<streamoff>(length_of_GDS));
+
+    // ------------------ Section 3 (BMS) ------------------
```
Same as above - move all sections into separate functions

ArtemiiH commented on main.cpp:
```diff
@@ -0,0 +1,216 @@
+        file.seekg(section3_start + static_cast<streamoff>(length_of_BMS));
+    }
+
+    // ------------------ Section 4 (BDS) ------------------
```
Same as above - move all sections into separate functions

ArtemiiH commented on main.cpp:
```diff
@@ -0,0 +1,216 @@
+        fout << (unsigned int)(unsigned char)data_bytes[i] << " ";
+    fout << endl;
+
+    // ------------------ GRIB message length check ------------------
```
Same as above - move all sections into separate functions

ArtemiiH commented on main.cpp:
```diff
@@ -0,0 +1,216 @@
+    fout << "Declared GRIB message length: " << length_from_GRIB << endl;
+    fout << "Sum of section lengths:       " << total_section_length << endl;
+
+    fout.close();
```
Also I would suggest, to separate file write functionality into a separate function as well. Something, that will receive file handle and all the structs from the previous functions (PDS, GDS, etc.)

ArtemiiH reviewed (commented):
Here are some comments, generally, you should refactor the code - split it up into functions, that will be much more readable, also, separate writing to the file and processing data.

ArtemiiH commented on main.cpp:
```diff
@@ -0,0 +1,216 @@
+    {
+        float f;
+        unsigned char b[4];
+    } ref;
```
It would be nice to also have more descriptive names of existing unions/structs/variables

ArtemiiH commented on main.cpp:
```diff
@@ -0,0 +1,216 @@
+    int hour = (unsigned char)datetime[3];
+    int minute = (unsigned char)datetime[4];
+
+    file.seekg(7, ios::cur); // Skip 7 bytes
```
Generally, such comments are useless, I can already see that you skip 7 bytes from the code, but it would be nice, if you could write the reason in the comment

Felix754 committed:
Version 2.0

-Renamed main.cpp to grib_decoder.cpp for convenience
-Removed the sample output file
-Moved section parsing and writing into separate functions for better readability
-Refactored the code into a CLI application
-Added several safety checks
-Improved the structure of comments alongside the code
-Moved the created functions and structures from grib_decoder.cpp to grib_sections.cpp

Felix754 committed:
Version 2.1

- Improved file structure to separate function access between user and developer code (grib_sections.cpp, grib_sections.hpp)
- Added option to run the program with command-line arguments
- Minor changes in function structure

Felix754 committed:
Delete OUT.txt | Version 2.1
